xarray/core/dataarray.py

```diff
@@ -422,9 +423,6 @@ def __init__(
     ) -> None:
         if fastpath:
             variable = data
-            assert dims is None
-            assert attrs is None
-            assert indexes is not None
```

Comment:
What happens when you set values other than None here now? Is there a test for this?

Reply:
i understood that `fastpath` indiciated:

* I know what i'm doing, please avoid using expensive type checking.

Do we need more safety measures? 

I feel like runtime checks should not need to be added to appease a linter. I under there is an environment variable that makes `assert` into a no-op, but I feel like we can avoid these lines of code if we can.

Reply:
Ok i added back the asserts. they did not have a significant effect on the benchmark I was running.

Reply:
I ended up reverting the changes to this method. thanks for the review.

Reply:
(sorry I made the changes on a local branch and forgot to push)

Reply:
I agree with you on the performance cost of `isinstance`. But if you want fast and correct code it's mypy that you should heavily rely on. Mypy won't be useful however if using overrides like `cast` and `type:ignore`. 

While reading up on the array api standard I've come to prefer the approach that `Array` classes should have the _exact_ required type for the input arguments, and use `xp.asarray` to handle all the other wider types. This way we separate the concerns in a familiar way and it significantly reduces the number of required `isinstance` checks internally because we can trust that the arguments always are correct. 

`DataArray` unfortunately took the other route when these kinds of performance bottlenecks wasn't as big of a priority and now we have to deal with the backwards compatibility concerns.